From b101837b27cfdc95cad4fb65e4a63388189d7f68 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 06 Jan 2021 14:10:31 +0200
Subject: [PATCH] [SW-0] bitops: introduce dedicated header

As a preparation for the Gaudi HBM code addition, we need to introduce
the ffs() function for finding set bits.
This already gives a rationale for a dedicated bitops header.
Introduce such a header and move all the relevant bit macros to the new
header.
This also remove the duplication of the EXTRACT_BITS() macro.

Change-Id: If5b3c4c302c4683a3abd8e45b8544c341d112332
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/spi.c b/common/spi.c
index d8f4b09..ea78d28 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -10,7 +10,8 @@
 #include <psoc_gpio3_regs.h>
 #include <psoc_spi_regs.h>
 
-#include "preboot.h"
+#include <bitops.h>
+#include <preboot.h>
 #include "common.h"
 #include <spi.h>
 #include "uart.h"
diff --git a/gaudi/asic.c b/gaudi/asic.c
index a41d695..2c77591 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -16,6 +16,7 @@
 #include <tpc1_qm_regs.h>
 
 #include <string.h>
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/gaudi/common.c b/gaudi/common.c
index 22366de..036650e 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -24,6 +24,7 @@
 #include <gaudi/gaudi_reg_map.h>
 #include <common/pci_ids.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/gaudi/reset.c b/gaudi/reset.c
index 81020c2..f0e54a6 100644
--- a/gaudi/reset.c
+++ b/gaudi/reset.c
@@ -47,6 +47,7 @@
 #include <tpc0_cfg_masks.h>
 #include <nic0_qm0_masks.h>
 
+#include <bitops.h>
 #include <preboot.h>
 
 #include "common.h"
diff --git a/gaudi/security.c b/gaudi/security.c
index 49f1c01..e0c47d0 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -46,6 +46,7 @@
 #include <nic0_txe0_masks.h>
 #include <nic0_tmr_masks.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <uart.h>
 #include <hl_boot_if.h>
diff --git a/gaudi2/asic.c b/gaudi2/asic.c
index 260ed5b..a7e7715 100644
--- a/gaudi2/asic.c
+++ b/gaudi2/asic.c
@@ -9,6 +9,7 @@
 #include <psoc_gpio0_regs.h>
 
 #include <string.h>
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/gaudi2/common.c b/gaudi2/common.c
index e9dd60d..d40514a 100644
--- a/gaudi2/common.c
+++ b/gaudi2/common.c
@@ -26,6 +26,7 @@
 #include <hl_boot_if.h>
 #include <gaudi2/gaudi2_reg_map.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/gaudi2/common.h b/gaudi2/common.h
index 4da82c2..ce05871 100644
--- a/gaudi2/common.h
+++ b/gaudi2/common.h
@@ -7,6 +7,7 @@
 #define COMMON_GAUDI2_H_
 
 #include <gaudi2_blocks_linux_driver.h>
+#include <bitops.h>
 
 #define HL_PCIE_LINK_RETRIES             50
 #define HL_FW_BOOT_PTRS_BASE		0x80000
diff --git a/goya/asic.c b/goya/asic.c
index 0d2f2bd..9bd1e54 100644
--- a/goya/asic.c
+++ b/goya/asic.c
@@ -8,6 +8,7 @@
 #include <goya_blocks.h>
 #include <psoc_global_conf_regs.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya/common.c b/goya/common.c
index be98799..ad62b51 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -16,6 +16,7 @@
 
 #include <cpu_timestamp_regs.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2/asic.c b/goya2/asic.c
index 8499ff4..0bd1d21 100644
--- a/goya2/asic.c
+++ b/goya2/asic.c
@@ -9,6 +9,7 @@
 #include <psoc_gpio0_regs.h>
 
 #include <string.h>
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2/common.c b/goya2/common.c
index 25a07ca..2a22d2e 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -29,6 +29,7 @@
 #include <dcon0_tstd_regs.h>
 #include "preboot/thermal.h"
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2/common.h b/goya2/common.h
index 9ae9adb..4fa1b72 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -8,6 +8,7 @@
 
 #include <goya2_blocks_linux_driver.h>
 #include <goya2/goya2.h>
+#include <bitops.h>
 
 #define PPBOOT_SIZE			0x4000		/* 16KB */
 #define PREBOOT_SIZE			0xC000		/* 48KB */
diff --git a/goya2_arc/asic.c b/goya2_arc/asic.c
index 4c7d680..c2f48c0 100644
--- a/goya2_arc/asic.c
+++ b/goya2_arc/asic.c
@@ -6,6 +6,7 @@
 
 #include <psoc_global_conf_regs.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2_arc/common.c b/goya2_arc/common.c
index aec5afc..782db03 100644
--- a/goya2_arc/common.c
+++ b/goya2_arc/common.c
@@ -22,6 +22,7 @@
 #include <common/pci_ids.h>
 #include <arcboot_export.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2_arc/common.h b/goya2_arc/common.h
index 2c8cc9f..8d5c415 100644
--- a/goya2_arc/common.h
+++ b/goya2_arc/common.h
@@ -8,6 +8,7 @@
 #define COMMON_GOYA2_ARC_H_
 
 #include <goya2_blocks_linux_driver.h>
+#include <bitops.h>
 
 #define PPBOOT_SIZE		0x4000		/* 16KB */
 #define PREBOOT_SIZE		0xC000		/* 48KB */
@@ -46,12 +47,5 @@
 
 #define ARC_RST_VALUE		0x110
 #define ARC_RST_VEC_ADDR	0x0
-
-/*
- * extracts "num" number of bits from position "pos" in the
- * given number "value"
- */
-#define EXTRACT_BITS(value, pos, num) ((value & (((1 << num) - 1) << pos))\
-						>> pos)
 
 #endif /* COMMON_GOYA2_ARC_H_ */
diff --git a/goya2_arc/pll.c b/goya2_arc/pll.c
index ea7f16a..ac448f7 100644
--- a/goya2_arc/pll.c
+++ b/goya2_arc/pll.c
@@ -5,6 +5,7 @@
  */
 
 #include <stddef.h>
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include "common.h"
diff --git a/goya2_arc_gnu/asic.c b/goya2_arc_gnu/asic.c
index cb41d80..fcc3647 100644
--- a/goya2_arc_gnu/asic.c
+++ b/goya2_arc_gnu/asic.c
@@ -16,6 +16,7 @@
 #include <dcon3_hbw_rtr_if0_rtr_h3_regs.h>
 #include <dcon3_hbw_rtr_if1_rtr_h3_regs.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2_arc_gnu/common.c b/goya2_arc_gnu/common.c
index e32fa31..7d520be 100644
--- a/goya2_arc_gnu/common.c
+++ b/goya2_arc_gnu/common.c
@@ -26,6 +26,7 @@
 #include <common/pci_ids.h>
 #include <goya2_blocks_linux_driver.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2_arc_gnu/common.h b/goya2_arc_gnu/common.h
index 63ce27f..dcef4fd 100644
--- a/goya2_arc_gnu/common.h
+++ b/goya2_arc_gnu/common.h
@@ -43,11 +43,4 @@
 #define HL_SPI_BASE		0x1A000000
 #define HL_CFG_BASE		0xFC000000
 
-/*
- * extracts "num" number of bits from position "pos" in the
- * given number "value"
- */
-#define EXTRACT_BITS(value, pos, num) ((value & (((1 << num) - 1) << pos))\
-						>> pos)
-
 #endif /* COMMON_GOYA2_ARC_GNU_H_ */
diff --git a/goya2_arc_gnu/pll.c b/goya2_arc_gnu/pll.c
index 8fe6040..7e69cfc 100644
--- a/goya2_arc_gnu/pll.c
+++ b/goya2_arc_gnu/pll.c
@@ -5,6 +5,7 @@
  */
 
 #include <stddef.h>
+#include <bitops.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
diff --git a/goya2_arc_gnu/security.c b/goya2_arc_gnu/security.c
index d389a2f..f57388e 100644
--- a/goya2_arc_gnu/security.c
+++ b/goya2_arc_gnu/security.c
@@ -5,6 +5,7 @@
  *
  */
 
+#include <bitops.h>
 #include <preboot.h>
 #include <string.h>
 #include <uart.h>
diff --git a/include/bitops.h b/include/bitops.h
new file mode 100644
index 0000000..ba2d559
--- /dev/null
+++ b/include/bitops.h
@@ -0,0 +1,56 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+#ifndef _BITOPS_H_
+#define _BITOPS_H_
+
+#define BIT(nr)		(1 << (nr))
+
+/* extracts "n" number of bits from position "p" in the given number "v" */
+#define EXTRACT_BITS(v, p, n) (((v) & (((1 << (n)) - 1) << (p))) >> (p))
+
+
+#ifndef __ASSEMBLY__
+
+/*
+ * __ffs32: find first bit set. This is defined the same way as
+ * the libc and compiler builtin ffs routines, therefore
+ * differs in spirit from the above ffz (man ffs).
+ */
+
+static inline int __ffs32(unsigned int x)
+{
+	int r = 1;
+
+	if (!x)
+		return 0;
+
+	if (!(x & 0xffff)) {
+		x >>= 16;
+		r += 16;
+	}
+	if (!(x & 0xff)) {
+		x >>= 8;
+		r += 8;
+	}
+	if (!(x & 0xf)) {
+		x >>= 4;
+		r += 4;
+	}
+	if (!(x & 3)) {
+		x >>= 2;
+		r += 2;
+	}
+	if (!(x & 1)) {
+		x >>= 1;
+		r += 1;
+	}
+	return r;
+}
+
+#endif /* __ASSEMBLY__ */
+
+#endif /* _BITOPS_H_ */
diff --git a/include/preboot.h b/include/preboot.h
index 2b80c9e..d7ac1ce 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -10,7 +10,6 @@
 
 #include <common.h>
 
-#define BIT(nr)					(1 << (nr))
 #define ARRAY_SIZE(x)				(sizeof(x) / sizeof((x)[0]))
 
 #ifdef __arc__
diff --git a/include/spi.h b/include/spi.h
index 49ded97..3f5b267 100644
--- a/include/spi.h
+++ b/include/spi.h
@@ -8,6 +8,8 @@
 #ifndef _HL_SPI_H_
 #define _HL_SPI_H_
 
+#include <bitops.h>
+
 /* The Status Register (SR) in use bits definition */
 #define READ_STATUS_WIP			BIT(0)
 #define READ_STATUS_BP0			BIT(2)
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index d2a011d..4c63b71 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -11,6 +11,7 @@
 #include <cpu_ca53_cfg_regs.h>
 #include <hl_fw_if.h>
 #include <hl_boot_if.h>
+#include <bitops.h>
 #include <preboot.h>
 
 #include "common.h"
diff --git a/preboot/i2cslave_indirect.c b/preboot/i2cslave_indirect.c
index c1b8fd8..b297529 100644
--- a/preboot/i2cslave_indirect.c
+++ b/preboot/i2cslave_indirect.c
@@ -7,6 +7,7 @@
 #include <stdint.h>
 #include <stdbool.h>
 #include <string.h>
+#include <bitops.h>
 #include <preboot.h>
 #include <uart.h>
 #include <common/module_i2c_slave.h>
diff --git a/preboot/module_pci.c b/preboot/module_pci.c
index 1e7b0e8..4f767c3 100644
--- a/preboot/module_pci.c
+++ b/preboot/module_pci.c
@@ -7,6 +7,7 @@
 #include <pcie_aux_regs.h>
 #include <pcie_core_regs.h>
 
+#include <bitops.h>
 #include <preboot.h>
 
 #include "cmd_module.h"
diff --git a/preboot/preboot.c b/preboot/preboot.c
index ca625f7..7326bbf 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -10,6 +10,7 @@
 
 #include <psoc_global_conf_regs.h>
 
+#include <bitops.h>
 #include <preboot.h>
 #include <hl_fw_if.h>
 #include <hl_boot_if.h>
